include/exec/__detail/__system_context_if.h

```diff
@@ -21,84 +21,121 @@
-  int,    // completion type: 0 for normal completion, 1 for cancellation, 2 for exception
-  void*); // If completion type is 2, this is the exception pointer.
-
-/// Callback to be called by the scheduler for each bulk item.
-typedef void (*__exec_system_context_bulk_item_callback_t)(
-  void*,          // data pointer passed to scheduler
-  unsigned long); // the index of the work item that is starting
-
-struct __exec_system_scheduler_interface {
-  /// The forward progress guarantee of the scheduler.
-  ///
-  /// 0 == concurrent, 1 == parallel, 2 == weakly_parallel
-  uint32_t __forward_progress_guarantee;
-
-  /// The size of the operation state object on the implementation side.
-  uint32_t __schedule_operation_size;
-  /// The alignment of the operation state object on the implementation side.
-  uint32_t __schedule_operation_alignment;
-
-  /// Schedules new work on the system scheduler, calling `cb` with `data` when the work can start.
-  /// Returns an object that should be passed to __destruct_schedule_operation when the operation completes.
-  void* (*__schedule)(
-    struct __exec_system_scheduler_interface* /*self*/,
-    void* /*__preallocated*/,
-    uint32_t /*__psize*/,
-    __exec_system_context_completion_callback_t /*cb*/,
-    void* /*data*/);
-
-  /// Destructs the operation state object.
-  void (*__destruct_schedule_operation)(
-    struct __exec_system_scheduler_interface* /*self*/,
-    void* /*operation*/);
-
-  /// The size of the operation state object on the implementation side.
-  uint32_t __bulk_schedule_operation_size;
-  /// The alignment of the operation state object on the implementation side.
-  uint32_t __bulk_schedule_operation_alignment;
-
-  /// Schedules new bulk work of size `size` on the system scheduler, calling `cb_item` with `data`
-  /// for indices in [0, `size`), and calling `cb` on general completion.
-  /// Returns the operation state object that should be passed to __destruct_bulk_schedule_operation.
-  void* (*__bulk_schedule)(
-    struct __exec_system_scheduler_interface* /*self*/,
-    void* /*__preallocated*/,
-    uint32_t /*__psize*/,
-    __exec_system_context_completion_callback_t /*cb*/,
-    __exec_system_context_bulk_item_callback_t /*cb_item*/,
-    void* /*data*/,
-    unsigned long /*size*/);
-
-  /// Destructs the operation state object for a bulk_schedule.
-  void (*__destruct_bulk_schedule_operation)(
-    struct __exec_system_scheduler_interface* /*self*/,
-    void* /*operation*/);
-};
+  struct system_context_interface;
+  struct system_scheduler_interface;
+
+  union __system_context_interface_reserved {
+    void* __object_ptr;
+    void* (*__function_ptr)();
+    uintptr_t __int;
+  };
+
+  /// Interface that allows interaction with the system context, allowing scheduling work on the system.
+  struct system_context_interface {
+    /// The supported version of the system context interface, in the form YYYYMM.
+    uint32_t version;
```

Comment:
It's not clear to me how this version number is planned to be used.

What will the client do if it sees a backend implementation that has a newer version than the version currently supported? Are all future versions guaranteed to be backwards compatible with older versions?
What if the client sees a backend implementation that is older that the current version? Will it need to introduce additional branches everywhere to check the version on every field access for fields which were added after the first version?

Reply:
> Are all future versions guaranteed to be backwards compatible with older versions?

yes, so that client code compiled against version N can still work with a system context from version N+1.

> What if the client sees a backend implementation that is older that the current version? Will it need to introduce additional branches everywhere to check the version on every field access for fields which were added after the first version?

Yes, new client code needs to do a version number check before using any field that was added after v1.